File: py/patterns/components/doors.py
from .common import MapSite
from .rooms import Room
from .spells import Spell


class Door(MapSite):
    def __init__(self, r1: Room, r2: Room):
        self.__room1 = r1
        self.__room2 = r2

    def enter(self):
        pass

    def get_room1(self):
        return self.__room1

    def get_room2(self):
        return self.__room2

    def __str__(self) -> str:
        return 'simple door between room {} and room {}'.format(self.__room1.get_number(), self.__room2.get_number())


class DoorNeedingSpell(Door):
    def __init__(self, r1: Room, r2: Room, spell: Spell):
        super().__init__(r1, r2)
        self.__spell = spell

    def __str__(self) -> str:
        return ('spell door between room {} and room {} with spell {}'
                .format(self.get_room1().get_number(), self.get_room2().get_number(), self.__spell.word()))

File: goimp/components/doors.go
package components

import "fmt"

type Door interface {
	MapSite
}

type SimpleDoor struct {
	room1  Room
	room2  Room
	isOpen bool
}

func CreateSimpleDoor(r1, r2 Room) Door {
	return &SimpleDoor{room1: r1, room2: r2}
}

func createSimpleDoor(r1, r2 Room) *SimpleDoor {
	return &SimpleDoor{room1: r1, room2: r2}
}

func (d SimpleDoor) Enter() {
	fmt.Printf("entering room 1: %T from room 2: %T, isOpen: %v", d.room1, d.room2, d.isOpen)
}

func (d SimpleDoor) String() string {
	return fmt.Sprintf("Simple Door{room 1: %T, room 2: %T, isOpen: %v}", d.room1, d.room2, d.isOpen)
}

type DoorNeedingSpell struct {
	room1  Room
	room2  Room
	isOpen bool
	spell  Spell
}

func CreateSpellDoor(r1, r2 Room) Door {
	return &DoorNeedingSpell{room1: r1, room2: r2, spell: Spell{"big magic~"}}
}

func createSpellDoor(r1, r2 Room) *DoorNeedingSpell {
	return &DoorNeedingSpell{room1: r1, room2: r2, spell: Spell{"big magic~"}}
}

func (d DoorNeedingSpell) Enter() {
	fmt.Printf("entering room 1: %T from room 2: %T, isOpen: %v", d.room1, d.room2, d.isOpen)
}

func (d DoorNeedingSpell) String() string {
	return fmt.Sprintf("Spell Door{room 1: %T, room 2: %T, isOpen: %v}", d.room1, d.room2, d.isOpen)
}
